x86-32 passing tests regression

Several tests failed at the same time for PR #7691 
(https://github.com/DynamoRIO/dynamorio/actions/runs/18724532585/job/53405532740?pr=7691)
```
code_api|tool.drcacheoff.burst_replaceall 
code_api|tool.drcacheoff.burst_syscall_inject 
code_api|tool.drcacheoff.legacy 
code_api|tool.drcacheoff.func_view_noret 
code_api|tool.drcacheoff.altbindir 
```

On top of:
```
code_api|client.drx-scattergather 
code_api|client.drx-scattergather-bbdup 
code_api|sample.memval_simple_scattergather 
code_api|tool.drcachesim.scattergather-x86 
```
Already reported in #7678 .

`code_api|tool.drcacheoff.burst_replaceall `: postcmd failed with segfault.
```
2025-10-22T17:36:10.1804922Z 371: Test timeout computed to be: 90
2025-10-22T17:36:10.1806749Z 371: Running |/opt/hostedtoolcache/cmake/3.19.7/x64/cmake-3.19.7-Linux-x86_64/bin/cmake;-E;remove_directory /home/runner/work/dynamorio/dynamorio/build_debug-internal-32/suite/tests/drmemtrace.tool.drcacheoff.burst_replaceall.27341.9864.dir|
2025-10-22T17:36:10.1809149Z 371: Running cmd |/home/runner/work/dynamorio/dynamorio/build_debug-internal-32/clients/bin32/tool.drcacheoff.burst_replaceall|
drmemtrace.tool.drcacheoff.burst_replace.27339.6707.dir|
2025-10-22T17:36:10.1815790Z 371: Running postcmd |/home/runner/work/dynamorio/dynamorio/build_debug-internal-32/clients/bin32/drmemtrace_launcher;-indir;/home/runner/work/dynamorio/dynamorio/build_debug-internal-32/suite/tests/drmemtrace.tool.drcacheoff.burst_replaceall.27368.5318.dir;-tool;basic_counts|
2025-10-22T17:36:10.1818587Z 371: CMake Error at /home/runner/work/dynamorio/dynamorio/suite/tests/process_cmdline.cmake:109 (message):
2025-10-22T17:36:10.1819609Z 371:   *** postcmd failed (Segmentation fault): ***
2025-10-22T17:36:10.1820191Z 371: 
2025-10-22T17:36:10.1820951Z 371: Call Stack (most recent call first):
2025-10-22T17:36:10.1821863Z 371:   /home/runner/work/dynamorio/dynamorio/suite/tests/runmulti.cmake:60 (process_cmdline)
2025-10-22T17:36:10.1822688Z 371: 
2025-10-22T17:36:10.1823208Z 371: 
2025-10-22T17:36:10.1823927Z       Test #371: code_api|tool.drcacheoff.burst_replaceall ........................***Failed    0.76 sec
```

`code_api|tool.drcacheoff.burst_syscall_inject`: trace invariant failure.
```
2025-10-22T17:36:10.9025126Z 373: CMake Error at /home/runner/work/dynamorio/dynamorio/suite/tests/process_cmdline.cmake:109 (message):
2025-10-22T17:36:10.9026092Z 373:   *** cmd failed (Subprocess aborted): Collecting a trace...
2025-10-22T17:36:10.9026590Z 373: 
2025-10-22T17:36:10.9026884Z 373:   Done collecting trace
2025-10-22T17:36:10.9027242Z 373: 
2025-10-22T17:36:10.9027605Z 373:   Testing system call trace template injection...
2025-10-22T17:36:10.9028077Z 373: 
2025-10-22T17:36:10.9028529Z 373:   Trace invariant failure in T27411 at ref # 100 at instruction # 49 (1
2025-10-22T17:36:10.9029335Z 373:   instrs since timestamp 152329120): Kernel trace-end branch marker does not
2025-10-22T17:36:10.9029956Z 373:   match next pc
2025-10-22T17:36:10.9030272Z 373: 
2025-10-22T17:36:10.9030528Z 373:   ***
2025-10-22T17:36:10.9030800Z 373: 
2025-10-22T17:36:10.9031102Z 373: Call Stack (most recent call first):
2025-10-22T17:36:10.9031879Z 373:   /home/runner/work/dynamorio/dynamorio/suite/tests/runmulti.cmake:57 (process_cmdline)
2025-10-22T17:36:10.9032619Z 373: 
2025-10-22T17:36:10.9032950Z 373: 
2025-10-22T17:36:10.9033570Z 371/469 Test #373: code_api|tool.drcacheoff.burst_syscall_inject ....................***Failed    0.35 sec
```

`code_api|tool.drcacheoff.legacy`: couldn't open trace file.
```
2025-10-22T17:37:43.4870066Z 459: Test timeout computed to be: 120
2025-10-22T17:37:43.4870438Z 459: Unexpected trace sequence
2025-10-22T17:37:43.4871456Z 459: Failed to initialize scheduler: Failed to open /home/runner/work/dynamorio/dynamorio/clients/drcachesim/tests/offline-legacy-trace.gz (was RLIMIT_NOFILE exceeded?)
2025-10-22T17:37:43.4872512Z 459: ERROR: failed to initialize analyzer
2025-10-22T17:37:43.4873396Z       Test #459: code_api|tool.drcacheoff.legacy ..................................***Failed  Required regular expression not found. Regex=[^
```

`tool.drcacheoff.func_view_noret`: couldn't open trace file.
```
2025-10-22T17:37:43.5172008Z         Start 460: code_api|tool.drcacheoff.func_view_noret
2025-10-22T17:37:43.5172317Z 
2025-10-22T17:37:43.5188607Z 460: Test command: /home/runner/work/dynamorio/dynamorio/build_debug-internal-32/bin32/runstats "-s" "90" "-killpg" "-silent" "-env" "LD_LIBRARY_PATH" "/home/runner/work/dynamorio/dynamorio/build_debug-internal-32/lib32/debug:/home/runner/work/dynamorio/dynamorio/build_debug-internal-32/ext/lib32/debug:" "-env" "DYNAMORIO_OPTIONS" "-stderr_mask 0xC -dumpcore_mask 0 -code_api" "/home/runner/work/dynamorio/dynamorio/build_debug-internal-32/clients/bin32/drmemtrace_launcher" "-indir" "/home/runner/work/dynamorio/dynamorio/clients/drcachesim/tests/offline-func-trace" "-tool" "func_view"
2025-10-22T17:37:43.5191871Z 460: Test timeout computed to be: 120
2025-10-22T17:37:43.5192244Z 460: Unexpected trace sequence
2025-10-22T17:37:43.5193644Z 460: Failed to initialize scheduler: Failed to open /home/runner/work/dynamorio/dynamorio/clients/drcachesim/tests/offline-func-trace/trace/drmemtrace.tool.fib_plus.125254.3691.trace.gz (was RLIMIT_NOFILE exceeded?)
2025-10-22T17:37:43.5194956Z 460: ERROR: failed to initialize analyzer
2025-10-22T17:37:43.5195887Z       Test #460: code_api|tool.drcacheoff.func_view_noret .........................***Failed  Required regular expression not found.
```

`code_api|tool.drcacheoff.altbindir`: trying to read an AARCH64 trace in x86 tests???
```
2025-10-22T17:37:43.5606796Z 461: CMake Error at /home/runner/work/dynamorio/dynamorio/suite/tests/runmulti.cmake:72 (message):
2025-10-22T17:37:43.5607758Z 461:   output |ERROR: failed to initialize analyzer: Directory setup failed:
2025-10-22T17:37:43.5608416Z 461:   Failed sanity checks for thread log file
2025-10-22T17:37:43.5609638Z 461:   /home/runner/work/dynamorio/dynamorio/build_debug-internal-32/suite/tests/drmemtrace.altbindir.aarch64/raw/drmemtrace.threadsig.16274.6003.raw.gz:
2025-10-22T17:37:43.5611069Z 461:   Architecture mismatch: trace recorded on aarch64 but tools built for i386
2025-10-22T17:37:43.5611699Z 461: 
2025-10-22T17:37:43.5612169Z 461:   | failed to match expected output |ERROR: failed to initialize analyzer:
2025-10-22T17:37:43.5613140Z 461:   Failed to create analysis tool: Tool failed to initialize: Failed to load
2025-10-22T17:37:43.5613915Z 461:   binaries: Failed to map module /tmp/nonexistent/threadsig
2025-10-22T17:37:43.5614459Z 461: 
2025-10-22T17:37:43.5614714Z 461:   |
2025-10-22T17:37:43.5614986Z 461: 
2025-10-22T17:37:43.5615248Z 461: 
2025-10-22T17:37:43.5616032Z       Test #461: code_api|tool.drcacheoff.altbindir ...............................***Failed    0.01 sec
```

--- a/clients/drcachesim/tests/offline-burst_syscall_inject.template
+++ b/clients/drcachesim/tests/offline-burst_syscall_inject.template
@@ -12,7 +12,11 @@ Output format:
 ------------------------------------------------------------
            1           0:       W0.T1 <marker: version 7>
 #ifdef X86
+# ifdef X64
            2           0:       W0.T1 <marker: filetype 0x4240>
+# else
+           2           0:       W0.T1 <marker: filetype 0x4220>
+# endif
 #else
            2           0:       W0.T1 <marker: filetype 0x4208>
 #endif
@@ -20,6 +24,7 @@ Output format:
            4           0:       W0.T1 <marker: page size 4096>
            5           0:       W0.T1 <marker: timestamp 1>
 #ifdef X86
+# ifdef X64
            6           0:       W0.T1 <marker: trace start for system call number 324>
            7           1:       W0.T1 ifetch       1 byte\(s\) @ 0x00000000deadbe00 90                   nop
            8           2:       W0.T1 ifetch       2 byte\(s\) @ 0x00000000deadbe01 0f 07                sysret \(target 0xf00d\)
@@ -32,6 +37,20 @@ Output format:
           15           4:       W0.T1 <marker: trace start for system call number 268435455>
           16           5:       W0.T1 ifetch       2 byte\(s\) @ 0x00000000f00d8b00 0f 07                sysret \(target 0xf00d\)
           17           5:       W0.T1 <marker: trace end for system call number 268435455>
+# else
+           6           0:       W0.T1 <marker: trace start for system call number 375>
+           7           1:       W0.T1 ifetch       1 byte\(s\) @ 0xdeadbe00 90                   nop
+           8           2:       W0.T1 ifetch       2 byte\(s\) @ 0xdeadbe01 0f 07                sysret \(target 0xf00d\)
+           9           2:       W0.T1 <marker: trace end for system call number 375>
+          10           2:       W0.T1 <marker: trace start for system call number 224>
+          11           3:       W0.T1 ifetch       2 byte\(s\) @ 0x8badf000 8b 12                mov    \(%edx\), %edx
+          12           3:       W0.T1 read         4 byte\(s\) @ 0xdecafbad by PC 0x8badf000
+          13           4:       W0.T1 ifetch       2 byte\(s\) @ 0x8badf002 0f 07                sysret \(target 0xf00d\)
+          14           4:       W0.T1 <marker: trace end for system call number 224>
+          15           4:       W0.T1 <marker: trace start for system call number 268435455>
+          16           5:       W0.T1 ifetch       2 byte\(s\) @ 0xf00d8b00 0f 07                sysret \(target 0xf00d\)
+          17           5:       W0.T1 <marker: trace end for system call number 268435455>
+# endif
 #else
            6           0:       W0.T1 <marker: trace start for system call number 283>
            7           1:       W0.T1 ifetch       4 byte\(s\) @ 0x00000000deadbe00 d503201f   nop

--- a/clients/drcachesim/tools/invariant_checker.cpp
+++ b/clients/drcachesim/tools/invariant_checker.cpp
@@ -1814,10 +1814,13 @@ invariant_checker_t::check_for_pc_discontinuity(
     if (shard->prev_kernel_end_branch_target_ != 0) {
         const addr_t kernel_end_branch_target = shard->prev_kernel_end_branch_target_;
         shard->prev_kernel_end_branch_target_ = 0;
-        if (kernel_end_branch_target != cur_pc)
+        if (kernel_end_branch_target != cur_pc &&
+            // We always see a discontinuity for sysenter.
+            prev_instr.instr.type != TRACE_TYPE_INSTR_SYSENTER) {
             // We want the check for PC continuity of cur_pc to take precedence
             // over this, so later code may overwrite error_msg.
             error_msg = "Kernel trace-end branch marker does not match next pc";
+        }
     }
     if (prev_instr_trace_pc == 0 /*first*/) {
         return error_msg;

--- a/suite/tests/CMakeLists.txt
+++ b/suite/tests/CMakeLists.txt
@@ -4444,7 +4444,8 @@ if (BUILD_CLIENTS)
     endif ()
 
     # Test reading a legacy pre-interleaved file in thread-sharded mode.
-    if (ZLIB_FOUND)
+    # The checked-in file is for x86_64 so we have to limit the test to that.
+    if (ZLIB_FOUND AND X64)
       torunonly_api(tool.drcacheoff.legacy "${drcachesim_path}" "offline-legacy.c" ""
         "-infile;${PROJECT_SOURCE_DIR}/clients/drcachesim/tests/offline-legacy-trace.gz;-no_core_sharded"
         OFF OFF)
@@ -4810,7 +4811,8 @@ if (BUILD_CLIENTS)
         "-use_physical" "@${test_mode_flag}@-tool@view" "")
     endif ()
 
-    if (UNIX AND ZLIB_FOUND)
+    # The checked-in file is for x86_64 so we have to limit the test to that.
+    if (UNIX AND ZLIB_FOUND AND X64)
       add_exe(tool.fib_plus
         ${PROJECT_SOURCE_DIR}/clients/drcachesim/tests/fib_plus.c)
       link_with_pthread(tool.fib_plus)
@@ -4922,10 +4924,15 @@ if (BUILD_CLIENTS)
       set(tool.drcacheoff.burst_replace_nodr ON)
       torunonly_drcacheoff(burst_replace tool.drcacheoff.burst_replace "" "" "")
 
-      set(tool.drcacheoff.burst_replaceall_nodr ON)
-      torunonly_drcacheoff(burst_replaceall tool.drcacheoff.burst_replaceall ""
-        "@-tool@basic_counts" "")
-      unset(tool.drcacheoff.burst_replaceall_rawtemp) # use preprocessor
+      # This test has not been updated to write out encodings.bin and so it has
+      # trouble with some vdso code in 32-bit. 32-bit is a low priority target
+      # so we do not plan to address this and simply disable for 32-bit.
+      if (X64)
+        set(tool.drcacheoff.burst_replaceall_nodr ON)
+        torunonly_drcacheoff(burst_replaceall tool.drcacheoff.burst_replaceall ""
+          "@-tool@basic_counts" "")
+        unset(tool.drcacheoff.burst_replaceall_rawtemp) # use preprocessor
+      endif ()
 
       if (X64 AND UNIX)
         set(tool.drcacheoff.burst_noreach_nodr ON)
@@ -5471,7 +5478,8 @@ if (BUILD_CLIENTS)
   # Limited to UNIX because the checked-in modules.log is in UNIX format
   # (Windows needs extra fields per library).
   # Limited to ZLIB_FOUND because we test gzipped .raw files here as well.
-  if (UNIX AND ZLIB_FOUND)
+  # The checked-in file is for x86_64 so we have to limit the test to that.
+  if (UNIX AND ZLIB_FOUND AND X64)
     # We make a writable copy so we can both write the final trace there and run our
     # analyzer in a single command line  Since we're testing postprocessing we
     # need to dynamically clobber the directory.

--- a/suite/tests/CMakeLists.txt
+++ b/suite/tests/CMakeLists.txt
@@ -5478,30 +5478,33 @@ if (BUILD_CLIENTS)
   # Limited to UNIX because the checked-in modules.log is in UNIX format
   # (Windows needs extra fields per library).
   # Limited to ZLIB_FOUND because we test gzipped .raw files here as well.
-  # The checked-in file is for x86_64 so we have to limit the test to that.
-  if (UNIX AND ZLIB_FOUND AND X64)
-    # We make a writable copy so we can both write the final trace there and run our
-    # analyzer in a single command line  Since we're testing postprocessing we
-    # need to dynamically clobber the directory.
-    set(srcdir
-      ${PROJECT_SOURCE_DIR}/clients/drcachesim/tests/drmemtrace.threadsig.aarch64.raw)
-    set(locdir ${CMAKE_CURRENT_BINARY_DIR}/drmemtrace.altbindir.aarch64)
-    file(MAKE_DIRECTORY ${locdir})
-    file(COPY ${srcdir}/raw DESTINATION ${locdir}/)
-    torunonly_api(tool.drcacheoff.altbindir "${drcachesim_path}"
-      "offline-altbindir.c" ""
-      "-indir;${locdir};-tool;opcode_mix;-alt_module_dir;${srcdir};-module_file;${locdir}/raw/modules.log"
-      OFF OFF)
-    set(tool.drcacheoff.altbindir_basedir
-      "${PROJECT_SOURCE_DIR}/clients/drcachesim/tests")
-    set(tool.drcacheoff.altbindir_runcmp "${CMAKE_CURRENT_SOURCE_DIR}/runmulti.cmake")
-    set(tool.drcacheoff.altbindir_precmd
+  if (UNIX AND ZLIB_FOUND)
+    # The checked-in file is for x86_64 so we have to limit the test to that.
+    if (X64)
+      # We make a writable copy so we can both write the final trace there and run our
+      # analyzer in a single command line  Since we're testing postprocessing we
+      # need to dynamically clobber the directory.
+      set(srcdir
+        ${PROJECT_SOURCE_DIR}/clients/drcachesim/tests/drmemtrace.threadsig.aarch64.raw)
+      set(locdir ${CMAKE_CURRENT_BINARY_DIR}/drmemtrace.altbindir.aarch64)
+      file(MAKE_DIRECTORY ${locdir})
+      file(COPY ${srcdir}/raw DESTINATION ${locdir}/)
+      torunonly_api(tool.drcacheoff.altbindir "${drcachesim_path}"
+        "offline-altbindir.c" ""
+        "-indir;${locdir};-tool;opcode_mix;-alt_module_dir;${srcdir};-module_file;${locdir}/raw/modules.log"
+        OFF OFF)
+      set(tool.drcacheoff.altbindir_basedir
+        "${PROJECT_SOURCE_DIR}/clients/drcachesim/tests")
+      set(tool.drcacheoff.altbindir_runcmp "${CMAKE_CURRENT_SOURCE_DIR}/runmulti.cmake")
+      set(tool.drcacheoff.altbindir_precmd
         "${CMAKE_COMMAND}@-E@remove_directory@${locdir}/trace")
-    set(tool.drcacheoff.altbindir_failok ON)
+      set(tool.drcacheoff.altbindir_failok ON)
+    endif ()
 
     # Test the legacy trace versions OFFLINE_FILE_VERSION_KERNEL_INT_PC-1 and
     # TRACE_ENTRY_VERSION_NO_KERNEL_PC.  This requires a checked-in trace and thus
     # -alt_module_dir; it has gzipped .raw files; hence here next to the altbindir test.
+    # This does support 32-bit by expecting an error message there.
     set(srcdir
       ${PROJECT_SOURCE_DIR}/clients/drcachesim/tests/drmemtrace.legacy-int-offs.raw)
     set(locdir ${PROJECT_BINARY_DIR}/drmemtrace.legacy-int-offs)

